### PR TITLE
Mul for MatZq and MatZ

### PR DESCRIPTION
**Pull request by Marcel583**
**Description**

This PR implements multiplication for MatZq and MatZ.

**Testing**

- [x] I added basic working examples (possibly in doc-comment)
- [x] I added tests for large (pointer representation) values
- [x] I triggered all possible errors in my test in every possible way
- [x] I included tests for all reasonable edge cases

**Checklist:**

- [x] I have performed a self-review of my own code
  - [x] The code provides good readability and maintainability s.t. it fulfills best practices like talking code, modularity, ...
    - [x] The chosen implementation is not more complex than it has to be
  - [x] My code should work as intended and no side effects occur (e.g. memory leaks)
  - [x] The doc comments fit our style guide
  - [x] I have credited related sources if needed

### CODE REVIEW

File: src/integer_mod_q/mat_zq/arithmetic/mul.rs [hunk 1]
```diff
@@ -51,6 +53,55 @@ impl Mul for &MatZq {
...
+    fn mul(self, other: &MatZ) -> Self::Output {
+        if self.get_num_columns() != other.get_num_rows() {
+            panic!("Tried to multiply matrices with mismatching matrix dimensions.");
+        }
+
+        let other_matzq = MatZq::from((other, &self.get_mod()));
+        let mut new =
+            MatZq::new(self.get_num_rows(), other.get_num_columns(), self.get_mod()).unwrap();
+        unsafe { fmpz_mod_mat_mul(&mut new.matrix, &self.matrix, &other_matzq.matrix) };
+        new
```
Comment (Marvin-Beckmann) [inline src/integer_mod_q/mat_zq/arithmetic/mul.rs:94]: This way we do not have to copy the MatZ. Consider adding reduce by a separate function (as I proposed in a different Kanboard task), but if not, I would be fine with my proposal as well.
```suggestion
        let mut new =
            MatZq::new(self.get_num_rows(), other.get_num_columns(), self.get_mod()).unwrap();
        unsafe {
            fmpz_mat_mul(&mut new.matrix.mat[0], &self.matrix.mat[0], &other.matrix);
            _fmpz_mod_mat_reduce(&mut new.matrix)
        }
        new
```

File: src/integer_mod_q/mat_zq/arithmetic/mul.rs [hunk 2]
```diff
@@ -51,6 +53,55 @@ impl Mul for &MatZq {
...
+    ///
+    /// let c = &a * &b;
+    /// let d = a * b;
+    /// let e = &c * d;
+    /// let f = c * &e;
+    /// ```
+    ///
+    /// # Errors and Failures
+    /// - Panics if the dimensions of `self` and `other` do not match for multiplication.
+    fn mul(self, other: &MatZ) -> Self::Output {
```
Comment (Marvin-Beckmann) [inline src/integer_mod_q/mat_zq/arithmetic/mul.rs:85]: Previously we always added mul_safe as a function that does not panic. Is there any reason you did it differently?

**Review by Marvin-Beckmann — CHANGES_REQUESTED** [review]
Generally looks good, just some changes for efficiency and a general question

File: src/integer_mod_q/mat_zq/arithmetic/mul.rs (hunk 2 above)
Comment (Marcel583) [reply src/integer_mod_q/mat_zq/arithmetic/mul.rs:85]: We would need mul_safe for both directions (or a macro for doing so). I thought that this might be unnecessary. But if you think we could need it, I can add it as well.

**Review by Marvin-Beckmann — APPROVED** [review]
